File: engineering/concurrency/13-singleflight/skeleton/python/main.py
# 并发模型：threading
import threading

class Call:
    def __init__(self):
        # TODO: 添加条件变量、结果、错误、完成标记
        pass

class Group:
    def __init__(self):
        self._mu = threading.Lock()
        self._m = {}

    def do(self, key, fn):
        with self._mu:
            if key in self._m:
                c = self._m[key]
                # TODO: 释放锁后等待 c 完成，返回共享结果
                return "", None
            c = Call()
            self._m[key] = c

        # TODO: 在持有 c 的锁或不持有全局锁的情况下执行 fn
        # TODO: 保存结果、标记完成、唤醒所有等待者
        # TODO: 从 self._m 中删除 key
        return "", None

def main():
    g = Group()
    exec_count = [0]
    mu = threading.Lock()
    n = 100
    threads = []

    def task():
        def fn():
            with mu:
                exec_count[0] += 1
            return "result"
        v, err = g.do("key", fn)
        if err is not None or v != "result":
            print(f"FAIL: unexpected result {v!r} err={err}")

    for _ in range(n):
        t = threading.Thread(target=task)
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    if exec_count[0] == 1:
        print(f"PASS: {n} concurrent requests deduped to {exec_count[0]} execution")
    else:
        print(f"FAIL: exec_count={exec_count[0]} (expected 1)")

if __name__ == "__main__":
    main()

File: engineering/concurrency/13-singleflight/skeleton/go/main.go
package main

import (
	"fmt"
	"sync"
	"sync/atomic"
)

// call 代表一个进行中的调用。
type call struct {
	// TODO: 添加结果、错误、是否完成、等待 channel
}

// Group singleflight 组。
type Group struct {
	mu sync.Mutex
	m  map[string]*call
}

// Do 执行 fn(key)，保证相同 key 的并发调用只执行一次 fn。
func (g *Group) Do(key string, fn func() (string, error)) (string, error) {
	g.mu.Lock()
	if g.m == nil {
		g.m = make(map[string]*call)
	}
	if c, ok := g.m[key]; ok {
		g.mu.Unlock()
		// TODO: 等待 c 完成，返回共享结果
		_ = c
		return "", nil
	}
	c := &call{}
	g.m[key] = c
	g.mu.Unlock()

	// TODO: 执行 fn，保存结果，标记完成，唤醒所有等待者
	// TODO: 从 map 中删除 key（调用完成后清理）
	return "", nil
}

func main() {
	var g Group
	var execCount int64
	const n = 100

	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			v, err := g.Do("key", func() (string, error) {
				atomic.AddInt64(&execCount, 1)
				return "result", nil
			})
			if err != nil || v != "result" {
				fmt.Printf("FAIL: unexpected result %q err=%v\n", v, err)
			}
		}()
	}
	wg.Wait()

	if execCount == 1 {
		fmt.Printf("PASS: %d concurrent requests deduped to %d execution\n", n, execCount)
	} else {
		fmt.Printf("FAIL: execCount=%d (expected 1)\n", execCount)
	}
}
